### ISSUE TITLE
Two different actions (for 2 different events)... different parameters, possible?

### ISSUE THREAD

**Issue by gremo** [hotwired/stimulus#553]
1. Form controller: `form_controller.js` `POST` data to the server using AJAX, emitting `form:submit_success` event on success, `form:submit_error` otherwise
2. GTM controller `gtm_controller.js` send data to the Google Tag Manager server.

How can I listen to two events using `data-action` attribute, and have different parameters for each event?

Right now, this is working perfectly fine, and I'm happy to have both controllers decoupled:

```html
<form data-controller="form gtm"
  data-action="form:submit_success->gtm#push"
  data-gtm-event-param="formContactSuccess"
  data-gtm-conversion-value-param="25">
</form>
```

But what if I want 2 actions, with different parameters?

- `form:submit_success->gtm#push` with `{ event: "formContactSuccess"  }`
- `form:submit_error->gtm#push` with `{ event: "formContactError"  }`

If not possible (as I guess), can you suggest any good method to to the same?

**Comment by marcoroth** [comment]
Hey @gremo

Yeah, I think what you are describing is not possible, unless you dynamically change the data attribute of the param in the `form` controller before you invoke the `gtm` controller action.

I guess the best way to approach this is to have two different controller actions in your `gtm` controller and to leave out the param entirely. 

One action for pushing the success event and one for pushing the error event.

```html
<form data-controller="form gtm"
  data-action="
    form:submit_success->gtm#pushSuccess 
    form:submit_error->gtm#pushError
  "
  data-gtm-conversion-value-param="25">
</form>
```

You can still have the `push` action - maybe even mark it as private - so you don't have to implement the logic twice and call it from the other two actions.

```js
// gtm_controller.js

// ...

  push(eventType) {
    // ...
  }

  pushSuccess() {
    this.push("formContactSuccess")
  }

  pushError() {
    this.push("formContactError")
  }
```

If you don't want to hardcode your event name strings in the controller you could add data attributes on the element (or use two Stimulus params).

Something like:
```html
<form data-controller="form gtm"  
  ...
  data-gtm-success-event="formContactSuccess"
  data-gtm-error-event="formContactError"
  ...
>
```

And then access them via `this.element.dataset.gtmSuccessEvent` and `this.element.dataset.gtmErrorEvent`

**Comment by lomby92** [comment]
@marcoroth 👍🏻👍🏻

Quoting from DOC:
> name your action methods based on what will happen when they’re called

See: https://stimulus.hotwired.dev/reference/actions#naming-conventions

**Comment by gremo** [comment]
Thanks for the ideas @marcoroth , you solution could work but have some downsides.

```html
<form data-controller="form gtm"
  data-action="
    form:submit_success->gtm#pushSuccess 
    form:submit_error->gtm#pushError
  "
  data-gtm-conversion-value-param="25">
</form>
```

In this case you are force to create 1 method for each event you have to push. This is not reusable, as another app could use a different event name or rename the `formContactSuccess`.  With `dataset ` same problem, you are locked into a fixed number of event names.

**Comment by marcoroth** [comment]
That's right. 

Maybe you might want to add the event name you want to push to the event detail in the `this.dispatch(...)` call in your `form` controller then and extract it that way. You will have that event and it's detail accessible in the `utm` controller action as the first argument

**Comment by gremo** [comment]
@marcoroth Sure, that's a good solution. And you're perfectly right about the `<body>`, my `gtm_controller.js` is attached to it and is listening to thoese events, pushing data to GTM.

Drawback of this solution is coupling the form to the `utm:push` event. Why the form should "know" about `utm:push`? Seems wrong to me. In the future, if another "action" should be taken after `form:submit_success`, I'll need to modify the form controller source code, not just the HTML.

I'm studying a more decoupled solution, involving an "event proxy", that's why the question in the first place:

1. **Form controller** emits `form:submit_success` and `form:submit_error`, knows nothing about GTM or what is going to happen after those events
2. The **GTM controller** sitting in the `<body>` is listening for `gtm:push` events
3. Finally the **event proxy** knows about both and it's listening to form events, "translating" those events into `gtm:push` events

But I guess this is not possible ATM with the current `params` implementation, at least in a "generic" way

**Comment by gremo** [comment]
I guess I cannot achieve what I'd like in the way I'd love to... so I'll close this issue.

**Comment by ricardograca** [comment]
I was having a similar issue and in my case it was because I had used `:prevent` action option on the first event. In that case the order is not the expected one. CodePen exemplifying the issue: https://codepen.io/devius1/pen/xxyqqLe?editors=1011

As a quick fix I just removed the `:prevent` and used good old `event.preventDefault()` on the action directly, but if this is the intended behavior it would be good if the documentation mentioned it.
